### DISCUSSION TITLE
How do evaluation and mutability work in Typst?

### DISCUSSION THREAD

**Discussion by timjs** [typst/typst#2501] [Q&A]
Moving a discussion from #2425 here to not pollute the issue.

Let me first and foremost say that Typst is a piece of technology I was waiting for. I've been looking into document systems for some time, and I really like how it works and what it can do!

I'm curious which definition of purity is used for Typst. Coming from functional programming, I use these two conditions:
* it is [referentially transparent](https://wiki.haskell.org/Referential_transparency): it does not depend on anything other than its parameters
* it has no [side effect](https://en.wikipedia.org/wiki/Side_effect_(computer_science)).

In my opinion, using `state` I can easily construct a function in Typst which does adhere to both conditions:
```typst
#let c = state("counter", 0)
#let step() = {
  c.update(x => x + 1)
  [State is now #c.display()]
}
#step()\
#step()\
```

`step` has definitely a side effect: it updates a "hidden" state, and it is not referential transparent, given the same parameters (none in this case) it produces different outputs...

@laurmaedje wrote in response:

> The `step` function itself is pure, both invocations of it return the same value. You could store the return value of `step()` in a variable and use it in place of the two individual `step()` calls and the result would be the same. What's impure is the runtime that renders the content resulting from your code. This is similar to IO Actions in Haskell.

I'm quite amazed! Writing this:
```typst
#let c = state("counter", 0)
#let step() = {
  c.update(x => x + 1)
  [State is now #c.display()]
}
#let r = step()
#r #r
```
indeed results in "State is now 1 State is now 2". So `r` kind of "saves the recipe (or computation) of the `step()` calculation which is evaluated by the runtime". But... if this would have been referentially transparent, I should be able to replace `r` with *the result of the recipe*, not the recipe itself. Performing `r` definitely has a side effect, so how can this be pure? 

Writing this out in Haskell using IO actions, would be:

```haskell
docIO = do
  -- Create a new reference
  c <- newIORef 0
  -- Write a "recipe" for a state action
  let step = do
        modifyIORef' c (+ 1)
        -- Get the value in the reference
        n <- readIORef c
        pure ("State is now " ++ show n)
  -- Action needs to be *performed* for the first time...
  x <- step
  -- ...and for the second time
  y <- step
  pure (x ++ "\n" ++ y)
```
State modifications and reads need to be performed in the IO monad (or ST monad if we only want state transformations). The `step` action needs to be performed twice *inside* the IO monad. So, when comparing `state` to Haskell IO actions, a Typst document can be seen as running *as a whole* inside a IO monad . Is that correct? But then, by definition and by observation, this is not pure. There is no pure Haskell function which can replicate the result from above.

Maybe another question to clear this up would be the following. Which extension to Typst, as it is now, would make it not pure any more?

**Comment by emilyyyylime** [comment]
I'm no expert in Haskall, but essentially consider how your Haskall code still has no side-effects, but simply stores into docIO (immutably!) all that is needed to perform the side effects when the program does eventually evaluate, just as you said happens with Typst 

Typst just simplifies this process a bit by unifying the syntax you use. But note that you can *never* look inside the result of `state.display()`, you can only wrap it (and apply set/show rules to it).

In fact, Typst's incremental compilation model is inherently incompatible with any impure code, because it has to be able to cache reproducable segments of the document and model the data flow to figure out what needs to be recompiled upon any change.

How I think of it (though I'm not sure exactly how accurate this really is), is that `state.display(f)` desugars into `locate(loc => f(state.at(loc)))`, and as you may or may not know, the result of `locate` is a context-aware opaque element that can only be fully realised in the rendering stages of the compilation, wherein it can 'look back' to find the most recent value the state has been given, and finally use the provided function to typeset it appropriately

**Reply by laurmaedje** [reply]
> How I think of it (though I'm not sure exactly how accurate this really is), is that state.display(f) desugars into locate(loc => f(state.at(loc)))

That's pretty accurate.

**Reply by emilyyyylime** [reply]
This is me being a bit pedantic but really the part I'm not so sure about is the 'looking back' to find the actual value of the state

**Reply by Enivex** [reply]
> > How I think of it (though I'm not sure exactly how accurate this really is), is that state.display(f) desugars into locate(loc => f(state.at(loc)))
> 
> That's pretty accurate.

Does that mean that if I want to display multiple states it would be more efficient to use locate manually so I can get all the state.at calls inside the same one?

**Reply by laurmaedje** [reply]
Maybe it's a bit more efficient. I wouldn't worry about it.

**Comment by timjs** [comment]
Ok, I experimented a bit more and the big point I missed here is that Typst seems to use a _call-by-name_ evaluation strategy. Is that correct @laurmaedje? Seems like the logical choice to make for text manipulation software! 

From the perspective of call-by-name, I can understand why a function like `state()` needs a key: it needs to know on which state cell it operates. That's also why leaving out the key, as @emilyyyylime suggested in #2425, could be problematic: binding such a call to a variable doesn't reduce it to a state cell address, like in conventional strict languages, it gets (hygienically) substituted as-is at every use site. Later on, conceptually after all substitutions happened, the typesetting engine evaluates the fully substituted content, while taking all state manipulations into account in the order it encounters them. 

With the memorisation Typst implements, this is maybe not how it is technically implemented, but it explains the difference between a "pure" substitution phase, and a "not so pure" evaluation phase. 

Are there any more subtleties like this about the scripting language which would be good to know?

**Reply by timjs** [reply]
@laurmaedje, could you confirm that the evaluation model of Typst is indeed call-by-name? And if there are more subtleties like this about the scripting language which would be good to know? Thanks in advance!

**Reply by Enivex** [reply]
> @laurmaedje, could you confirm that the evaluation model of Typst is indeed call-by-name? And if there are more subtleties like this about the scripting language which would be good to know? Thanks in advance!

If you don't get an answer here you might have better luck on Discord

**Reply by laurmaedje** [reply]
@timjs Typst's evaluation phase isn't call-by-name, it's call-by-value. However, your description about how content is handled is relatively accurate. So I guess the way content works is somewhat similar to call-by-name. A content value can be substituted in multiple places and the fully built content is then processed by the "backend" of Typst.

**Reply by timjs** [reply]
Sorry for the late reply!

I fully understood what your previous comment was about after reading the [design document in the repo](https://github.com/typst/typst/blob/main/docs/dev/architecture.md). You mentioned Typst has separate Evaluation and Layout phases (although Layout can trigger Evaluation again). After inspecting the type of `state.display()` and other methods, which clearly state (pun not intended) that they produce `content`, I realised how things work. Thanks for the discussion about this!

I also see that Typst uses call-by-value, which is observable when creating your own `or`-function for example:
```typst
#let x = 0
#let or-(a, b) = {a or b}
#{if x == 0 or 1/x > 0.3 [true] else [false]} // is ok
#{if or-(x == 0, 1/x > 0.3) [true] else [false]} // triggers error at or-
```
(Call-by-name/value/need are only indistinguishable in a pure and total language.)

I think that, as sketched above, with call-by-name one could lift `state` methods to the Evaluation layer; but implementing call-by-value is usually easier.
